### PR TITLE
Updates in  synchronization mutex for Shared data

### PR DESCRIPTION
**Pull request by Naviabheeman**
Changed CCriticalSections to Mutex and RecursiveMutex as in Bitcoin. 

Add GUARDED_BY and EXCLUSIVE_LOCK_REQUIRED annotations to some shared data only. 
  Adding them to all shared data and functions caused lots of build errors on macOS. Clang supports static analysis for thread safety with option -thread-safety-analysis. But this does not seem to be strong enough to identify scope. i.e. when lock is outside a class or a function where the data is accessed. As most of our data and locks are global, and locking is in the outer scope, it causes lots of false positives. Adding more LOCKS in the same scope to pass the compilation warning causes deadlock in our tests.

### CODE REVIEW

File: src/net.h [hunk 1]
```diff
@@ -373,8 +373,8 @@ class CConnman
     static bool NodeFullyConnected(const CNode* pnode);
 
     // Network usage totals
-    CCriticalSection cs_totalBytesRecv;
-    CCriticalSection cs_totalBytesSent;
+    Mutex cs_totalBytesRecv;
+    Mutex cs_totalBytesSent;
```
Comment (azuchi) [inline src/net.h:377]: In Bitcoin, RecursiveMutex is used.

File: src/net.h [hunk 2]
```diff
@@ -625,15 +625,15 @@ class CNode
     size_t nSendOffset; // offset inside the first vSendMsg already sent
     uint64_t nSendBytes;
     std::deque<std::vector<unsigned char>> vSendMsg;
-    CCriticalSection cs_vSend;
-    CCriticalSection cs_hSocket;
-    CCriticalSection cs_vRecv;
+    RecursiveMutex cs_vSend;
+    RecursiveMutex cs_hSocket;
+    RecursiveMutex cs_vRecv;
```
Comment (azuchi) [inline src/net.h:630]: In Bitcoin, Mutex is used.

File: src/net.h [hunk 3]
```diff
@@ -394,17 +394,17 @@ class CConnman
...
+    RecursiveMutex cs_setBanned;
     bool setBannedIsDirty;
     bool fAddressesInitialized;
     CAddrMan addrman;
     std::deque<std::string> vOneShots;
-    CCriticalSection cs_vOneShots;
+    Mutex cs_vOneShots;
     std::vector<std::string> vAddedNodes GUARDED_BY(cs_vAddedNodes);
-    CCriticalSection cs_vAddedNodes;
+    Mutex cs_vAddedNodes;
```
Comment (azuchi) [inline src/net.h:404]: In Bitcoin, RecursiveMutex is used.

File: src/net.h [hunk 4]
```diff
@@ -394,17 +394,17 @@ class CConnman
...
     std::atomic<bool> fNetworkActive;
     banmap_t setBanned;
-    CCriticalSection cs_setBanned;
+    RecursiveMutex cs_setBanned;
     bool setBannedIsDirty;
     bool fAddressesInitialized;
     CAddrMan addrman;
     std::deque<std::string> vOneShots;
-    CCriticalSection cs_vOneShots;
+    Mutex cs_vOneShots;
```
Comment (azuchi) [inline src/net.h:402]: In Bitcoin, RecursiveMutex is used.

File: src/net.h [hunk 5]
```diff
@@ -530,7 +530,7 @@ struct LocalServiceInfo {
     int nPort;
 };
 
-extern CCriticalSection cs_mapLocalHost;
+extern Mutex cs_mapLocalHost;
```
Comment (azuchi) [inline src/net.h:533]: In Bitcoin, RecursiveMutex is used.

File: src/net.h (hunk 1 above)
Comment (Naviabheeman) [reply src/net.h:377]: they have only one mutex. other is changed to atomic 
```    
    // Network usage totals
    mutable Mutex m_total_bytes_sent_mutex;
    std::atomic<uint64_t> nTotalBytesRecv{0};
    ```

File: src/net.h (hunk 4 above)
Comment (Naviabheeman) [reply src/net.h:402]: This is completely removed actually. Bitcoin has reimplemented the peer management layer.

File: src/net.h (hunk 3 above)
Comment (Naviabheeman) [reply src/net.h:404]: this remains the same:
```
mutable Mutex m_added_nodes_mutex;
```

File: src/net.h (hunk 5 above)
Comment (Naviabheeman) [reply src/net.h:533]:
```
extern GlobalMutex g_maplocalhost_mutex;
```
you mean they say globalmutex which is actually
```
class GlobalMutex : public Mutex { };
```

**Review by azuchi — APPROVED** [review]

**Pull request merged**